File: test/language/choice_types/python/EnumParamChoiceTest.py
import zserio

import ChoiceTypes


class EnumParamChoiceTest(ChoiceTypes.TestCase):
    def testConstructor(self):
        selector = self.api.Selector.BLACK
        enumParamChoice = self.api.EnumParamChoice(selector)
        self.assertEqual(selector, enumParamChoice.selector)

        selector = self.api.Selector.GREY
        enumParamChoice = self.api.EnumParamChoice(selector, grey_=1234)
        self.assertEqual(selector, enumParamChoice.selector)
        self.assertEqual(1234, enumParamChoice.grey)

    def testFromReader(self):
        selector = self.api.Selector.GREY
        value = 234
        writer = zserio.BitStreamWriter()
        self._writeEnumParamChoiceToStream(writer, selector, value)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        enumParamChoice = self.api.EnumParamChoice.from_reader(reader, selector)
        self.assertEqual(selector, enumParamChoice.selector)
        self.assertEqual(value, enumParamChoice.grey)

    def testEq(self):
        enumParamChoice1 = self.api.EnumParamChoice(self.api.Selector.BLACK)
        enumParamChoice2 = self.api.EnumParamChoice(self.api.Selector.BLACK)
        self.assertTrue(enumParamChoice1 == enumParamChoice2)

        value = 99
        enumParamChoice1.black = value
        self.assertFalse(enumParamChoice1 == enumParamChoice2)

        enumParamChoice2.black = value
        self.assertTrue(enumParamChoice1 == enumParamChoice2)

        diffValue = value + 1
        enumParamChoice2.black = diffValue
        self.assertFalse(enumParamChoice1 == enumParamChoice2)

    def testHash(self):
        enumParamChoice1 = self.api.EnumParamChoice(self.api.Selector.BLACK)
        enumParamChoice2 = self.api.EnumParamChoice(self.api.Selector.BLACK)
        self.assertEqual(hash(enumParamChoice1), hash(enumParamChoice2))

        value = 99
        enumParamChoice1.black = value
        self.assertTrue(hash(enumParamChoice1) != hash(enumParamChoice2))

        enumParamChoice2.black = value
        self.assertEqual(hash(enumParamChoice1), hash(enumParamChoice2))

        diffValue = value + 1
        enumParamChoice2.black = diffValue
        self.assertTrue(hash(enumParamChoice1) != hash(enumParamChoice2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(63073, enumParamChoice1.__hash__())
        self.assertEqual(63074, enumParamChoice2.__hash__())

    def testGetSelector(self):
        selector = self.api.Selector.BLACK
        enumParamChoice = self.api.EnumParamChoice(selector)
        self.assertEqual(selector, enumParamChoice.selector)

    def testGetSetBlack(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.BLACK)
        value = 99
        enumParamChoice.black = value
        self.assertEqual(value, enumParamChoice.black)

    def testGetSetGrey(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.GREY)
        value = 234
        enumParamChoice.grey = value
        self.assertEqual(value, enumParamChoice.grey)

    def testGetSetWhite(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.WHITE)
        value = 65535
        enumParamChoice.white = value
        self.assertEqual(value, enumParamChoice.white)

    def testChoiceTag(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.BLACK)
        self.assertEqual(enumParamChoice.CHOICE_BLACK, enumParamChoice.choice_tag)

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.GREY)
        self.assertEqual(enumParamChoice.CHOICE_GREY, enumParamChoice.choice_tag)

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.WHITE)
        self.assertEqual(enumParamChoice.CHOICE_WHITE, enumParamChoice.choice_tag)

    def testBitSizeOf(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.BLACK)
        self.assertEqual(8, enumParamChoice.bitsizeof())

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.GREY)
        self.assertEqual(16, enumParamChoice.bitsizeof())

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.WHITE)
        self.assertEqual(32, enumParamChoice.bitsizeof())

    def testInitializeOffsets(self):
        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.BLACK)
        bitPosition = 1
        self.assertEqual(9, enumParamChoice.initialize_offsets(bitPosition))

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.GREY)
        self.assertEqual(17, enumParamChoice.initialize_offsets(bitPosition))

        enumParamChoice = self.api.EnumParamChoice(self.api.Selector.WHITE)
        self.assertEqual(33, enumParamChoice.initialize_offsets(bitPosition))

    def testReadWrite(self):
        selector = self.api.Selector.BLACK
        enumParamChoice = self.api.EnumParamChoice(selector)
        byteValue = 99
        enumParamChoice.black = byteValue
        writer = zserio.BitStreamWriter()
        enumParamChoice.write(writer)
        readEnumParamChoice = self.api.EnumParamChoice(selector)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readEnumParamChoice.read(reader)
        self.assertEqual(byteValue, readEnumParamChoice.black)
        self.assertEqual(enumParamChoice, readEnumParamChoice)

        selector = self.api.Selector.GREY
        shortValue = 234
        enumParamChoice = self.api.EnumParamChoice(selector, grey_=shortValue)
        writer = zserio.BitStreamWriter()
        enumParamChoice.write(writer)
        readEnumParamChoice = self.api.EnumParamChoice(selector)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readEnumParamChoice.read(reader)
        self.assertEqual(shortValue, readEnumParamChoice.grey)
        self.assertEqual(enumParamChoice, readEnumParamChoice)

        selector = self.api.Selector.WHITE
        enumParamChoice = self.api.EnumParamChoice(selector)
        intValue = 65535
        enumParamChoice.white = intValue
        writer = zserio.BitStreamWriter()
        enumParamChoice.write(writer)
        readEnumParamChoice = self.api.EnumParamChoice(selector)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readEnumParamChoice.read(reader)
        self.assertEqual(intValue, readEnumParamChoice.white)
        self.assertEqual(enumParamChoice, readEnumParamChoice)

    def _writeEnumParamChoiceToStream(self, writer, selector, value):
        if selector == self.api.Selector.BLACK:
            writer.write_signed_bits(value, 8)
        elif selector == self.api.Selector.GREY:
            writer.write_signed_bits(value, 16)
        elif selector == self.api.Selector.WHITE:
            writer.write_signed_bits(value, 32)
        else:
            self.fail(f"Invalid selector: {selector}")

File: test/language/choice_types/cpp/EnumParamChoiceTest.cpp
#include "gtest/gtest.h"

#include "choice_types/enum_param_choice/Selector.h"
#include "choice_types/enum_param_choice/EnumParamChoice.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace choice_types
{
namespace enum_param_choice
{

class EnumParamChoiceTest : public ::testing::Test
{
protected:
    void writeEnumParamChoiceToByteArray(zserio::BitStreamWriter& writer, Selector selector, int32_t value)
    {
        switch (selector)
        {
        case Selector::BLACK:
            writer.writeSignedBits(value, 8);
            break;

        case Selector::GREY:
            writer.writeSignedBits(value, 16);
            break;

        case Selector::WHITE:
            writer.writeSignedBits(value, 32);
            break;

        default:
            FAIL() << "Bad choice selector";
        }
    }

    EnumParamChoice::ParameterExpressions parameterExpressionsBlack = {
            nullptr, 0, [](void*, size_t) { return Selector::BLACK; } };
    EnumParamChoice::ParameterExpressions parameterExpressionsGrey = {
            nullptr, 0, [](void*, size_t) { return Selector::GREY; } };
    EnumParamChoice::ParameterExpressions parameterExpressionsWhite = {
            nullptr, 0, [](void*, size_t) { return Selector::WHITE; } };

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

TEST_F(EnumParamChoiceTest, emptyConstructor)
{
    {
        EnumParamChoice enumParamChoice;
        ASSERT_THROW(enumParamChoice.getSelector(), zserio::CppRuntimeException);
    }
    {
        EnumParamChoice enumParamChoice = {};
        ASSERT_THROW(enumParamChoice.getSelector(), zserio::CppRuntimeException);
    }
}

TEST_F(EnumParamChoiceTest, bitStreamReaderConstructor)
{
    const int8_t value = 99;
    zserio::BitStreamWriter writer(bitBuffer);
    writeEnumParamChoiceToByteArray(writer, Selector::BLACK, value);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    EnumParamChoice enumParamChoice(reader, parameterExpressionsBlack);
    ASSERT_EQ(Selector::BLACK, enumParamChoice.getSelector());
    ASSERT_EQ(value, enumParamChoice.getBlack());
}

TEST_F(EnumParamChoiceTest, copyConstructor)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsBlack);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);

    const EnumParamChoice enumParamChoiceCopy(enumParamChoice);
    ASSERT_EQ(Selector::BLACK, enumParamChoiceCopy.getSelector());
    ASSERT_EQ(value, enumParamChoiceCopy.getBlack());
}

TEST_F(EnumParamChoiceTest, assignmentOperator)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsGrey);
    const int16_t value = 234;
    enumParamChoice.setGrey(value);

    EnumParamChoice enumParamChoiceCopy;
    enumParamChoiceCopy = enumParamChoice;
    ASSERT_EQ(Selector::GREY, enumParamChoiceCopy.getSelector());
    ASSERT_EQ(value, enumParamChoiceCopy.getGrey());
}

TEST_F(EnumParamChoiceTest, moveCopyConstructor)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsBlack);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);

    // note that it doesn't ensure that move ctor was called
    const EnumParamChoice enumParamChoiceMoved(std::move(enumParamChoice));
    ASSERT_EQ(Selector::BLACK, enumParamChoiceMoved.getSelector());
    ASSERT_EQ(value, enumParamChoiceMoved.getBlack());
}

TEST_F(EnumParamChoiceTest, moveAssignmentOperator)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsGrey);
    const int16_t value = 234;
    enumParamChoice.setGrey(value);

    // note that it doesn't ensure that move ctor was called
    EnumParamChoice enumParamChoiceMoved;
    enumParamChoiceMoved = std::move(enumParamChoice);
    ASSERT_EQ(Selector::GREY, enumParamChoiceMoved.getSelector());
    ASSERT_EQ(value, enumParamChoiceMoved.getGrey());
}

TEST_F(EnumParamChoiceTest, propagateAllocatorCopyConstructor)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsBlack);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);

    const EnumParamChoice enumParamChoiceCopy(zserio::PropagateAllocator, enumParamChoice,
            EnumParamChoice::allocator_type());
    ASSERT_EQ(Selector::BLACK, enumParamChoiceCopy.getSelector());
    ASSERT_EQ(value, enumParamChoiceCopy.getBlack());
}

TEST_F(EnumParamChoiceTest, initialize)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsGrey);
    ASSERT_EQ(Selector::GREY, enumParamChoice.getSelector());
}

TEST_F(EnumParamChoiceTest, getSelector)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsBlack);
    ASSERT_EQ(Selector::BLACK, enumParamChoice.getSelector());
}

TEST_F(EnumParamChoiceTest, getSetBlack)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsBlack);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);
    ASSERT_EQ(value, enumParamChoice.getBlack());
}

TEST_F(EnumParamChoiceTest, getSetGrey)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsGrey);
    const int16_t value = 234;
    enumParamChoice.setGrey(value);
    ASSERT_EQ(value, enumParamChoice.getGrey());
}

TEST_F(EnumParamChoiceTest, getSetWhite)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(parameterExpressionsWhite);
    const int32_t value = 65535;
    enumParamChoice.setWhite(value);
    ASSERT_EQ(value, enumParamChoice.getWhite());
}

TEST_F(EnumParamChoiceTest, choiceTag)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(parameterExpressionsBlack);
    ASSERT_EQ(EnumParamChoice::CHOICE_black, enumParamChoiceB.choiceTag());

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(parameterExpressionsGrey);
    ASSERT_EQ(EnumParamChoice::CHOICE_grey, enumParamChoiceG.choiceTag());

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(parameterExpressionsWhite);
    ASSERT_EQ(EnumParamChoice::CHOICE_white, enumParamChoiceW.choiceTag());
}

TEST_F(EnumParamChoiceTest, bitSizeOf)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(parameterExpressionsBlack);
    ASSERT_EQ(8, enumParamChoiceB.bitSizeOf());

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(parameterExpressionsGrey);
    ASSERT_EQ(16, enumParamChoiceG.bitSizeOf());

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(parameterExpressionsWhite);
    ASSERT_EQ(32, enumParamChoiceW.bitSizeOf());
}

TEST_F(EnumParamChoiceTest, initializeOffsets)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(parameterExpressionsBlack);
    const size_t bitPosition = 1;
    ASSERT_EQ(9, enumParamChoiceB.initializeOffsets(bitPosition));

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(parameterExpressionsGrey);
    ASSERT_EQ(17, enumParamChoiceG.initializeOffsets(bitPosition));

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(parameterExpressionsWhite);
    ASSERT_EQ(33, enumParamChoiceW.initializeOffsets(bitPosition));
}

TEST_F(EnumParamChoiceTest, operatorEquality)
{
    EnumParamChoice enumParamChoice1;
    enumParamChoice1.initialize(parameterExpressionsBlack);
    EnumParamChoice enumParamChoice2;
    enumParamChoice2.initialize(parameterExpressionsBlack);
    ASSERT_TRUE(enumParamChoice1 == enumParamChoice2);

    const int8_t value = 99;
    enumParamChoice1.setBlack(value);
    ASSERT_FALSE(enumParamChoice1 == enumParamChoice2);

    enumParamChoice2.setBlack(value);
    ASSERT_TRUE(enumParamChoice1 == enumParamChoice2);

    const int8_t diffValue = value + 1;
    enumParamChoice2.setBlack(diffValue);
    ASSERT_FALSE(enumParamChoice1 == enumParamChoice2);
}

TEST_F(EnumParamChoiceTest, hashCode)
{
    EnumParamChoice enumParamChoice1;
    enumParamChoice1.initialize(parameterExpressionsBlack);
    EnumParamChoice enumParamChoice2;
    enumParamChoice2.initialize(parameterExpressionsBlack);
    ASSERT_EQ(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    const int8_t value = 99;
    enumParamChoice1.setBlack(value);
    ASSERT_NE(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    enumParamChoice2.setBlack(value);
    ASSERT_EQ(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    const int8_t diffValue = value + 1;
    enumParamChoice2.setBlack(diffValue);
    ASSERT_NE(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    // use hardcoded values to check that the hash code is stable
    ASSERT_EQ(63073, enumParamChoice1.hashCode());
    ASSERT_EQ(63074, enumParamChoice2.hashCode());
}

TEST_F(EnumParamChoiceTest, write)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(parameterExpressionsBlack);
    const int8_t valueB = 99;
    enumParamChoiceB.setBlack(valueB);
    zserio::BitStreamWriter writerB(bitBuffer);
    enumParamChoiceB.write(writerB);

    zserio::BitStreamReader readerB(writerB.getWriteBuffer(), writerB.getBitPosition(), zserio::BitsTag());
    EnumParamChoice readEnumParamChoiceB(readerB, parameterExpressionsBlack);
    ASSERT_EQ(valueB, readEnumParamChoiceB.getBlack());

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(parameterExpressionsGrey);
    const int16_t valueG = 234;
    enumParamChoiceG.setGrey(valueG);
    zserio::BitStreamWriter writerG(bitBuffer);
    enumParamChoiceG.write(writerG);

    zserio::BitStreamReader readerG(writerG.getWriteBuffer(), writerG.getBitPosition(), zserio::BitsTag());
    EnumParamChoice readEnumParamChoiceG(readerG, parameterExpressionsGrey);
    ASSERT_EQ(valueG, readEnumParamChoiceG.getGrey());

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(parameterExpressionsWhite);
    const int32_t valueW = 65535;
    enumParamChoiceW.setWhite(valueW);
    zserio::BitStreamWriter writerW(bitBuffer);
    enumParamChoiceW.write(writerW);

    zserio::BitStreamReader readerW(writerW.getWriteBuffer(), writerW.getBitPosition(), zserio::BitsTag());
    EnumParamChoice readEnumParamChoiceW(readerW, parameterExpressionsWhite);
    ASSERT_EQ(valueW, readEnumParamChoiceW.getWhite());
}

} // namespace enum_param_choice
} // namespace choice_types
